source/non_matching/fe_immersed_values.cc

```diff
@@ -0,0 +1,184 @@
+    // DoFHandler used by this cell, are the same
+    Assert(static_cast<const FiniteElementData<dim> &>(*this->fe) ==
+             static_cast<const FiniteElementData<dim> &>(cell->get_fe()),
+           (typename FEValuesBase<dim>::ExcFEDontMatch()));
+
+    this->maybe_invalidate_previous_present_cell(cell);
+    this->present_cell = {cell};
+
+    // This was the part of the work that is dependent on the actual data type
+    // of the iterator. Now pass on to the function doing the real work.
+    do_reinit();
+  }
+
+
+
+  template <int dim>
+  void
+  FEImmersedSurfaceValues<dim>::do_reinit()
+  {
+    // First call the mapping and let it generate the data specific to the
+    // mapping.
+    if (this->update_flags & update_mapping)
+      {
+        this->get_mapping().fill_fe_immersed_surface_values(
+          this->present_cell,
+          quadrature,
+          *this->mapping_data,
+          this->mapping_output);
+      }
+
+    // Call the finite element and, with the data already filled by the mapping,
+    // let it compute the data for the mapped shape function values, gradients
+    // etc.
+    this->get_fe().fill_fe_values(this->present_cell,
+                                  CellSimilarity::none,
+                                  this->quadrature,
+                                  this->get_mapping(),
+                                  *this->mapping_data,
+                                  this->mapping_output,
+                                  *this->fe_data,
+                                  this->finite_element_output);
+  }
+
+
+
+  template <int dim>
+  Tensor<1, dim>
+  FEImmersedSurfaceValues<dim>::shape_surface_grad(
+    const unsigned int function_no,
+    const unsigned int quadrature_point) const
+  {
+    const unsigned int component = 0;
+    return shape_surface_grad_component(function_no,
+                                        quadrature_point,
+                                        component);
+  }
+
+
+
+  template <int dim>
+  Tensor<1, dim>
+  FEImmersedSurfaceValues<dim>::shape_surface_grad_component(
+    const unsigned int function_no,
+    const unsigned int quadrature_point,
+    const unsigned int component) const
+  {
+    const Tensor<1, dim> gradient =
+      this->shape_grad_component(function_no, quadrature_point, component);
+    const Tensor<1, dim> &normal = this->normal_vector(quadrature_point);
+
+    return gradient - (normal * gradient) * normal;
+  }
+
+
+
+  template <int dim>
+  const ImmersedSurfaceQuadrature<dim> &
+  FEImmersedSurfaceValues<dim>::get_quadrature() const
+  {
+    return quadrature;
+  }
+
+
+
+  template <int dim>
+  inline void
+  FEImmersedSurfaceValues<dim>::initialize(const UpdateFlags update_flags)
+  {
+    UpdateFlags flags = this->compute_update_flags(update_flags);
+
+    if (flags & (update_JxW_values | update_normal_vectors))
+      flags |= update_covariant_transformation;
```

Comment:
Wasn't really sure if this was the right place to set this flag.

Reply:
Where is `update_covariant_transformation` needed?

Reply:
When computing normals and the `JxW`-values in  `MappingQ::fill_fe_immersed_surface_values`. Line 1363 in `mapping_q.cc`

Reply:
Indeed. The covariant/contravariant transformations are the most fundamental flags to compute the Jacobians of the transformation between reference and real cell, and as such the necessary step before one can fill the JxW or normal vectors, which depend on the Jacobian.